### ISSUE TITLE
Trying to build on MacBook Air 2022 (M2)

### ISSUE THREAD

**Issue by tinker2much** [pwsafe/pwsafe#899]
I'm trying to build 1.15 on a new M2 MacBook Air, under MacOS Ventura with Xcode 14.1 and wxWidgets 3.2.1. I am neither an experienced Mac developer nor familiar with the pwsafe source - this is pure "do what worked before and hope for the best". You've been kind enough to help me before when I got stuck on Mac builds; I beg your indulgence again. The procedure noted below last worked for me in June on an old Intel Mac Mini Late 2012, under Catalina, with wxwidgets 3.1.6. I am happy to try anything you suggest, provide further information, whatever ...

20221207        try to build latest Apple Silicon pwsafe
        download 115 version of pwsafe (pwsafe-1.15.0.tar.gz) and copy to documents
        set up "pwsafe" as a link to it
        download latest wxwidgets (3.2.1) and copy it to Documents
        build debug version
                cd wxwidgets-3.2.1
                mkdir static-debug
                cd static-debug
                /Users/peter/Documents/pwsafe/Misc/osx-build-wx -d
                time make -j`nproc`
                (actually enough to peg the mac! - 1% idle)
        repeat for release
                cd wxwidgets-3.2.1
                mkdir static-release
                cd static-release
                /Users/peter/Documents/pwsafe/Misc/osx-build-wx
                time make -j`nproc`
                        make -j `nproc`  914.68s user 140.66s system 744% cpu 2:21.77 total

        Go to the Xcode directory in pwsafe sources, and do these
                ./generate-configs -d /Users/peter/Documents/wxWidgets-3.2.1/static-debug/wx-config > pwsafe-debug.xcconfig
                ./generate-configs -r /Users/peter/Documents/wxWidgets-3.2.1/static-release/wx-config > pwsafe-release.xcconfig

                open the project file(dir) from within pwsafw/Xcode...
                clean the build
                run the build
                build fails on pwsafe-cli, if not more ...
        >>>>>>>>>> stuck here <<<<<<<<<<

**Comment by tinker2much** [comment]
I pasted that block in as code, hope it formats better. 
`20221207        try to build latest Apple Silicon pwsafe
        download 115 version of pwsafe (pwsafe-1.15.0.tar.gz) and copy to documents
        set up "pwsafe" as a link to it
        download latest wxwidgets (3.2.1) and copy it to Documents
        build debug version
                cd wxwidgets-3.2.1
                mkdir static-debug
                cd static-debug
                /Users/peter/Documents/pwsafe/Misc/osx-build-wx -d
                time make -j`nproc`
                (actually enough to peg the mac! - 1% idle)
        repeat for release
                cd wxwidgets-3.2.1
                mkdir static-release
                cd static-release
                /Users/peter/Documents/pwsafe/Misc/osx-build-wx
                time make -j`nproc`
                        make -j `nproc`  914.68s user 140.66s system 744% cpu 2:21.77 total

        Go to the Xcode directory in pwsafe sources, and do these
                ./generate-configs -d /Users/peter/Documents/wxWidgets-3.2.1/static-debug/wx-config > pwsafe-debug.xcconfig
                ./generate-configs -r /Users/peter/Documents/wxWidgets-3.2.1/static-release/wx-config > pwsafe-release.xcconfig

                open the project file(dir) from within pwsafw/Xcode...
                clean the build
                run the build
                build fails on pwsafe-cli, if not more ...
        >>>>>>>>>> stuck here <<<<<<<<<<
`

**Comment by tinker2much** [comment]
Here are the errors I see: 

> pwsafe-cli
> Undefined symbol: wxGet_wxConvLibcPtr()
> 
> Undefined symbol: wxFileName::Assign(wxString const&, wxPathFormat)
> 
> Undefined symbol: wxArrayString::Init(bool)
> 
> Undefined symbol: wxArrayString::~wxArrayString()
> 
> Undefined symbol: wxTranslations::GetUntranslatedString(wxString const&)
> 
> Undefined symbol: wxTranslations::Get()
> 
> Undefined symbol: wxMimeTypesManager::GetFileTypeFromExtension(wxString const&)
> 
> Undefined symbol: wxMimeTypesManager::wxMimeTypesManager()
> 
> Undefined symbol: wxMimeTypesManager::~wxMimeTypesManager()
> 
> Undefined symbol: wxString::ConvertStr(char const*, unsigned long, wxMBConv const&)
> 
> Undefined symbol: wxString::npos
> 
> Undefined symbol: wxPrivate::GetUntypedNullData()
> 
> Undefined symbol: wxFileType::GetMimeType(wxString*) const
> 
> Undefined symbol: wxTranslations::GetTranslatedString(wxString const&, wxString const&, wxString const&) const
> 
> Undefined symbol: _wxConvLibcPtr
>

**Comment by ronys** [comment]
Looks like the Xcode mac build is missing the wxWidgets libraries in the cli link.
Sorry I can't be more specific, as I don't have access to a Mac to build this.
You might want to try building on Mac using CMake, as this is what runs on GitHub's CI pipeline.

**Comment by tinker2much** [comment]
A different attempt: I got the latest wxwidgets from homebrew (3.2.1 for Apple Silicon). I did a get pull on pwsafe so I should have the latest version; I'm on the master branch. I generated the configs in the pwsafe Xcode folder pointing to that new wx-config. I started xcode and cleaned the build folder and tried to run the build. See attached screenshot. Is there some confusion about arm64 versus z86_64 there? 

<img width="1168" alt="Screenshot 2022-12-17 at 1 29 54 PM" src="https://user-images.githubusercontent.com/57112204/208263333-74d8a9d3-e9da-4e03-a256-26b7a2b08221.png">

**Comment by ronys** [comment]
Yep, looks like wx was built for Arm (macOs-arm64), but pwsafe-cli is being built for macOS-86_64.
Linking these together is like mixing oil and water, a no-go.

**Comment by tinker2much** [comment]
Right, but I'm using the latest wxwidgets, and the latest pwsafe from git, therefore the latest pwsafe xcode project, so what else can I do? Is there something with the project that's not quite right yet for Apple Silicon/Ventura/Xcode14.2?
I would be happy to try anything you suggest. (Like a passenger being instructed how to land the 747...)

**Comment by tinker2much** [comment]
> Looks like the Xcode mac build is missing the wxWidgets libraries in the cli link. Sorry I can't be more specific, as I don't have access to a Mac to build this. You might want to try building on Mac using CMake, as this is what runs on GitHub's CI pipeline.

I tried that and got errors about not finding OPENSSL (which I *do* have installed via homebrew), but cmake kindly told me how set the OPENSSL_ROOT_DIR value before the cmake invocation. Then, as you see below, there's another error about magic not being found. I'm not sure how far down this rabbit hole I should go. 

<img width="1252" alt="Screenshot 2022-12-19 at 11 06 24 AM" src="https://user-images.githubusercontent.com/57112204/208481945-ed641d49-6eb7-4af5-b4cc-f8d1383277e8.png">

**Comment by ronys** [comment]
Try:
$ brew install libmagic
$ brew link libmagic  (if the link is already created is going to fail, don't worry about that)

**Comment by tinker2much** [comment]
Progress:
<img width="1257" alt="Screenshot 2022-12-19 at 11 47 49 AM" src="https://user-images.githubusercontent.com/57112204/208488993-8116c864-0f9c-4042-97ef-5c7dc443396f.png">

**Comment by tinker2much** [comment]
Do I need to install more dependencies, as in the script for Linux: 
Misc/setup-linux-dev-env.sh ?

**Comment by ronys** [comment]
Nice catch.

Either pull the latest changes from master, or change pwsqrencode.mm to QREncode.mm in src/ui/wxWidgets/CMakeLists.txt

**Comment by tinker2much** [comment]
Thanks you. I got the latest changes, also brew installed and linked xerces-c, and now the main app builds. That's all I need, but FYI, the cli build still fails, see below. I don't know where to obtain libuuid, I don't think its in homebrew:

> [ 88%] Linking CXX executable pwsafe-cli
> ld: library not found for -luuid
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> make[2]: *** [cli/pwsafe-cli] Error 1
> make[1]: *** [cli/CMakeFiles/pwsafe-cli.dir/all] Error 2
> make: *** [all] Error 2

**Comment by tinker2much** [comment]
> That's all I need, but FYI, the cli build still fails, see below

To clarify, I don't need the cli either from the make or Xcode builds, but I would like to be able to build the main app in Xcode. Which still fails, while on the cli, and I think that prevents it from getting to the main app.

**Comment by ronys** [comment]
Odd that libuuid can be found for the main app but not the cli, since it's needed by both. Can you build the tests (make test)?

**Comment by tinker2much** [comment]
> peter@MacBook-Air build % make test 2>&1 | tee maketest_out.txt
> Running tests...
> Test project /Users/peter/Documents/pwsafe-git/build
> No tests were found!!!

**Comment by tinker2much** [comment]
There's still these two problems, even after the cmake build finishes:

> peter@MacBook-Air ~ % pwsafe
> Could not initialize help subsystem.

and:
<img width="250" alt="Screenshot 2022-12-21 at 10 06 47 AM" src="https://user-images.githubusercontent.com/57112204/209035402-16fa8f75-b7e5-4449-a637-ebcc7b986b76.png">

**Comment by ronys** [comment]
Help files: cd to pwsafe/help in the source tree, run make, then copy the zip files that were created to /usr/share/passwordsafe/help (after creating that directory).
xsd files: Copy the contents of the pwsafe/xml directory to /usr/share/passwordsafe/xml.

**Comment by tinker2much** [comment]
> peter@MacBook-Air share % sudo mkdir passwordsafe
> Password:
> mkdir: passwordsafe: Operation not permitted
> Press any key to continue...                                                                                                                                         
> 
> peter@MacBook-Air share % sudo su
> peter@MacBook-Air share % sudo su
> sh-3.2# mkdir /usr/share/passwordsafe
> mkdir: /usr/share/passwordsafe: Operation not permitted
>

**Comment by ronys** [comment]
Hm. It seems that Apple prefers locally installed software to be under /usr/local, not /usr ([reference](https://apple.stackexchange.com/questions/301976/mac-os-x-wont-let-me-mkdir-usr-share-tomcat6-even-as-root)).
Try putting things there (e.g., /usr/local/bin, /usr/local/share, etc.).

**Comment by tinker2much** [comment]
I made those new directories under /usr/local instead and copied the xml and help files there. I copied the app to /usr/local/bin, but it still reports the same errors.

**Comment by ronys** [comment]
I think pwsafe needs to be "pointed" to those directories. How to do so requires more research than I can reasonable do without access to a Mac, but I'd either rebuild pwsafe with these directories replacing the defaults, or look for environment variables that can be used to override the /usr/ defaults with /usr/local.

**Comment by tinker2much** [comment]
I pulled the latest from GitHub, which included 1.16. I thought it was worth a try in case something changed or improved regarding my mac issue (I'm an optimist), so I retried cmake, and also the Xcode build, and both had the same problems as before with the cli portion (perhaps related to it being confused about x86_64 versus aarch64?)

As before, I'm happy to help in any way I can, mostly by trying anything you can suggest, as a kind of remote typist on hardware you don't otherwise have access to.

**Comment by tinker2much** [comment]
As another data point, I just built both wxwidgets 3.2.1 via make and the latest pwsafe via cmake, on x86_64 (in fact, a repurposed 2012 mac mini now under Debian Bullseye 11.6). No problems at all.

**Comment by nobugshere** [comment]
Hello, 
In case you're still having trouble with Xcode building for x86_64: I see a problem in your Xcode screenshot that I didn't see mentioned in this thread:

<img width="432" alt="x86build" src="https://user-images.githubusercontent.com/61946508/212500581-fa4b9403-d331-4f86-ad7a-0371ef4f0231.png">

"Rosetta" means it's set to compile for x86_64 and run under Rosetta.  If you want to build a binary for Rosetta (or a universal binary) you need to have x86_64 versions of all the dependencies, including the Apple frameworks that wxWidgits might use.  If you just want to build for your M1/M2 Mac, you need to click on that text and select the plain "My Mac" setting:

![Arm build](https://user-images.githubusercontent.com/61946508/212500606-9c9eb97b-d19e-42a6-99e2-092ad9a9a6c2.png)

For what it's worth, I've been building pwsafe in a similar method: pulling from git and using the homebrew wxWidgits build.  I haven't needed to install any other dependencies.  I've made a few tweaks to the Xcode deployment settings to install the "release" build to my ~/Applications folder and it works great, including the in-app help.  

My current system:
Mac Studio M1 Max
MacOS 13.1 Ventura
Xcode 14.2
wxWidgets 3.2.1 (installed via homebrew)
PasswordSafe 1.16 pulled from git master on 2023-01-03

I hope some of this helps!

**Comment by tinker2much** [comment]
I started from scratch and got things to build just fine. I'm pretty sure I had tried without the Rosetta choice, too, but after so many attempts who knows what exact combinations of things I had tried. Thanks for your help. 

I put together a script to do all of the steps I did, starting with obtaining pwsafe and wxwidgets, including any other preparatory steps mentioned in the README.MAC.DEVELOPERS.md document, through copying the resulting app to /Applications. You can also rerun it in future to pull the latest pwsafe and rebuild. Offered here, in case it helps anyone else. 

```
# build or rebuild pwsafe using homebrew wxWidgets and xcode on mac

cd ~/Documents

if [ ! -d "pwsafe/" ]; then
  brew install wxwidgets
  git clone --depth 1 https://github.com/pwsafe/pwsafe.git
fi

cd pwsafe

git stash push
git pull

# "You have to either touch version.wx or remove/rename version.h before build with Xcode"
touch ./version.wx
rm src/ui/wxWidgets/version.h

# For users using Xcode to build the application you must link to the wxWidgets libraries and include files. You need to build **pwsafe-release.xcconfig** and **pwsafe-debug.xcconfig** files in the **Xcode** directory, derived from **wx-config** file located in **static-release** and **static-debug** folder of the native generated wxWidgets environment (see above).
cd Xcode
./generate-configs -r /opt/homebrew/bin/wx-config > pwsafe-release.xcconfig
./generate-configs -d /opt/homebrew/bin/wx-config > pwsafe-debug.xcconfig

# open xcode, where as the user, clean the build directory and then build the app, then exit
#   -W means the open will wait until you exit to return
open -W pwsafe-xcode6.xcodeproj

git stash drop

# Find the latest copy of the app in some obscure place and copy it to Applications  
find ~ -name pwsafe.app  2>/dev/null | xargs ls -artd | tail -n1 | xargs -I HHEERREE cp -pR HHEERREE /Applications

```

**Comment by tinker2much** [comment]
I had gotten V1.16 to build on my MacBook Air M2 a year ago using the script I posted above; that doesn't work now for v1.18.2. 

I tried to follow the latest README.MAC.DEVELOPERS.md instructions, which mention that you now need yubikey (or maybe you don't).
1. The doc suggested following appropriate bits of the workflow at pwsafe/.github/workflows/macos-latest.yml to build the yubikey stuff. When I did so, I got errors about a missing "a2x" - a program associated with asciidoctor that doesn't seem to come with the home-brew asciidoctor. Stumped there. 
2. The doc also suggested "a *Debug-no-yubi* configuration that is the same as Debug", but I couldn't find that in the Xcode project. Stumped again. 

Has anyone succeeded at building the latest pwsafe on an M2 Mac?

Though I'm an experienced software developer, and I'm game to try, my expertise lies in entirely different areas. As I noted in the original post: "I am neither an experienced Mac developer nor familiar with the pwsafe source - this is pure "do what worked before and hope for the best". You've been kind enough to help me before when I got stuck on Mac builds; I beg your indulgence again."

**Comment by nobugshere** [comment]
Hi @tinker2much,
Much has changed since 1.16 :-)

1. When you ran the configure script for the Yibikey libs, did you include the --disable-documentation option?  That removes the need for a2x.  (I think you might still get the man pages.)
2. If you are using the xcodebuild command, as shown in the README, all you need to do is pass Debug-no-yubi as the parameter to -configuration.  If you are using the Xcode GUI: Select the menu item “Product->Scheme->Edit Scheme…”; Then, in the dialog window, select “Run” on the left side, then in the “Build Configuration” drop-down list, you should see the Debug-no-yubi option.

A few other points:
- I should mention I only tested the Debug-no-yubi configuration while using the pwsafe-debug scheme (Select that first from the Product->Scheme menu.)
- If you are still using wxWidgets 3.2.1, I recommend you update to at least 3.2.2.1. I seem to recall there was a problem with 3.2.1.  I've been using 3.2.4 with no problems, but the official release still uses 3.2.2.1.
- Your script implies using the homebrew wxWidgets.  That's fine, but you will need to check "Disable Library Validation" in Xcode, as described in the README.  The Hardened Runtime was added sometime after 1.16.

My system is an M1 Max running Sonoma 14.3, so it definitely can be done!

Good luck!
Richard
